### PR TITLE
Make it possible for `TTNT::TestSelector` to determine base sha and target sha by itself

### PR DESCRIPTION
**Pull request by Genki-S**
@robin850 could you review the idea and the code?
( :warning: I am not confident if my explanation is clear enough :sweat: So, if I'm not clear enough please tell me which point I should clarify more :pray: )

I introduced some rules in order to make the problem of finding base commit (to which diff should be calculated) much simpler. Currently I'm taking the approach to store the sha of the commit into a file ('.ttnt/commit_obj.txt'), but it's not elegant I think. Here are the rules:

- When `rake ttnt:test:anchor` is run and succeeded, test-to-code mapping file is committed automatically ( https://github.com/Genki-S/ttnt/commit/546861c3e8d82693f5d4ee9b7cbf29ac2c487e95 )
  - this makes "anchored commit" (the commit on which coverage information is recorded) and "anchoring commit" (the commit test-to-code mapping file is saved) atomic
- When `rake ttnt:test:anchor` is run, the working tree should be clean ( https://github.com/Genki-S/ttnt/commit/fb77e0cfb8ac53062856d7a1b349694efdb52eec )
  - this assures that anchoring commit does not include coverage information which cannot be reproduced on anchored commit

By these rules, when a user runs `rake ttnt:test:run`, we only have to do following steps to calculate which tests to run:

1. Regard 'HEAD' as the target commit. ( https://github.com/Genki-S/ttnt/commit/9540c95c8bf1d6b3d8cc51375bb37ec7794f3c30 )
1. Find the latest commit test-to-code mapping is saved. Regard this as the base commit. ( https://github.com/Genki-S/ttnt/commit/0427829a92fdfede972e7d30baceeddfbfb56430 )
1. Read the test-to-code mapping from the base commit ( https://github.com/Genki-S/ttnt/commit/9df58f9de65533232fa78850ab24c9d22f8c69cf )
1. Calculate the tests to run based on the test-to-code mapping and the diff between base commit and target commit ( this is the same as before and not changed in this PR )
1. Run selected tests ( this is not changed as well )

This makes it unneccessary to store the sha of commit `rake ttnt:test:anchor` has been run on. So I removed it (https://github.com/Genki-S/ttnt/commit/01dd0b96579d4f27e2687656b8e0cd67213b65b6 ).

### CODE REVIEW

**Comment by robin850** [comment]
Hi Genki!

Sorry for the late answer! I'm not sure whether committing when running `rake ttnt:test:anchor` is a good idea, for several reasons:

* During the process, the command might be run several times as new changes are added or things get removed.
* This will clearly crap the logs ; having once every two times a commit that's kind of useless inside the history is clearly not something that people want.
* The number of commits will uselessly grow ; on a project like Rails, where there's already 50K+ commits, having even more commits is a no-go I think.
* Thus, doing `git bisect` will be more painful as it will run a command for *each* commit between two points in the history.

I'm not sure to understand the second change. Why running anchoring in an uncommitted environment is a problem ? I think that this is most likely the moment when you can call it, before testing and committing.

Sorry if I'm missing something!

**Comment by Genki-S** [comment]
Hi Robin!

> Sorry for the late answer!

Please don't be. I always appreciate (and actually totally surprised by) your quick responses! Let me thank you for that :smile:

---

## About auto committing

> During the process, the command might be run several times as new changes are added or things get removed

I don't want users to run anchoring that frequently... My intention is like this:

1. a user merges a feature branch
2. a user runs `rake ttnt:test:anchor` to save current test-to-code mapping
3. a user starts another feature branch
4. a user runs `rake ttnt:test:run` and tests related to changes are selected and run

So, I want users to run `rake ttnt:test:anchor` once in a while, when they think `rake ttnt:test:run` becomes to produce many tests (meaning code has changed so hugely that many tests are affected and should be run).

But your other 3 points are so true... Especially I did not think of the effect to `git bisect`. I don't want developers to stop over auto-generated TTNT commits (they sure will hate me for that :P ). It sounds sticking to the approach of saving the sha of the commit `rake ttnt:test:anchor` is run is better for now (FYI, the downside is that users should be conscious of adding TTNT generated files themselves).

I'll remove the change of automatically making commits from this PR soon.

---

## About avoiding dirty working tree

> I'm not sure to understand the second change. Why running anchoring in an uncommitted environment is a problem ? I think that this is most likely the moment when you can call it, before testing and committing.

This is because our test selection depends on the diff in code history. Say, in the latest commit (ababab), we have the simplest 1 to 1 test-to-code mapping like this:

```
| test          |   | code     |
|---------------|---|----------|
| fizz-test     | - | fizz     |
| buzz-test     | - | buzz     |
| fizzbuzz-test | - | fizzbuzz |
```

(meaning, `fizz-test` executes the code portion `fizz`, `buzz-test` executes `buzz`, and so on)

And the user removed one mapping on working tree like this:

```
| test          |   | code     |
|---------------|---|----------|
| fizz-test     |   | fizz     |
| buzz-test     | - | buzz     |
| fizzbuzz-test | - | fizzbuzz |
```

(`fizz-test` no more executes `fizz`)

In this case when a user runs `rake ttnt:test:anchor`, the test coverage of **working tree** is saved. Now regard the user commits TTNT generated files.

Now, think that the user completely discards the working tree (`git reset --hard HEAD`) and changes `fizz` code and commit the change. Now, since `rake ttnt:test:anchor` was run on dirty working tree, it does not know `fizz` code affects `fizz-test`. But since the commit "ababab" maps `fizz-test` to `fizz` code, `fizz-test` is expected to be run.

In conclusion, saving the test-to-code mapping of dirty working tree leads to saving the mapping which is dependent on code which might not be on the git history. This is, I will call, a faulty mapping. (In one word, anchored commit should always be reachable from 'HEAD', and dirty working tree is not always reachable from 'HEAD'.)

I'm not sure if I could explain this clearly enough. Feel free to ask me more questions so that I can clear things up (in my mind as well).

**Comment by robin850** [comment]
> I always appreciate (and actually totally surprised by) your quick responses! Let me thank you for that

:smiley: :heart:

> FYI, the downside is that users should be conscious of adding TTNT generated files themselves

Yep it seems fine as it will be listed running `git status`. :-)

> Now, think that the user completely discards the working tree (`git reset --hard HEAD`) and changes `fizz` code and commit the change.

I guess you wrongly assumed in your reasoning that the previous change (i.e. committing when running `rake ttnt:test:anchor`) was still there. If the user runs `git reset`, then even the mapping file would be rolled back to the state it was on `HEAD`. Am I right ? Let me know if I misunderstand something.

**Comment by Genki-S** [comment]
> If the user runs git reset, then even the mapping file would be rolled back to the state it was on HEAD

Oh, that's right. I'm sorry my `git reset --hard HEAD` example was not good... but users can reset files except for TTNT generated files, so I think committing faulty mapping can still be possible.

However, I noticed that preventing anchoring on dirty working tree does not help preventing faulty mapping from being committed if commit cannot be automatically generated... My change here is meaningful if both of (1) preventing anchoring on dirty working tree and (2) committing TTNT files automatically are there.

I still think that current usage is a bit too much dependent on users' operations. So I'm gonna close this PR and think other approaches to alleviate users' operations as much as possible. Thank you for your feedback! :smile:
